File: patterns/01-design-patterns/02-structural/04-proxy/python/proxy.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class DocumentStore:
    def __init__(self):self.documents={"public":"status:ok","secret":"key:rotated"}
    def read(self,name):
        if name not in self.documents:raise PatternError("not_found")
        return self.documents[name]
class DocumentProxy:
    def __init__(self,role):self.role=role;self.subject=None;self.load_count=0
    def read(self,name):
        if name=="secret" and self.role!="admin":raise PatternError("forbidden")
        if self.subject is None:self.subject=DocumentStore();self.load_count+=1
        return self.subject.read(name)
def evaluate(input_data):
    proxy=DocumentProxy(input_data.get("role","viewer"));values=[proxy.read(name) for name in input_data.get("reads",[])];return{"values":values,"load_count":proxy.load_count}

File: patterns/01-design-patterns/02-structural/04-proxy/js/proxy.mjs
export class PatternError extends Error{constructor(code){super(code);this.code=code;}}
class DocumentStore{constructor(){this.documents=new Map([["public","status:ok"],["secret","key:rotated"]]);}read(name){if(!this.documents.has(name))throw new PatternError("not_found");return this.documents.get(name);}}
class DocumentProxy{constructor(role){this.role=role;this.subject=undefined;this.loadCount=0;}read(name){if(name==="secret"&&this.role!=="admin")throw new PatternError("forbidden");if(this.subject===undefined){this.subject=new DocumentStore();this.loadCount++;}return this.subject.read(name);}}
export function evaluate(input){const proxy=new DocumentProxy(input.role??"viewer");const values=(input.reads??[]).map((name)=>proxy.read(name));return{values,load_count:proxy.loadCount};}
